server/src/main/java/org/opensearch/cluster/routing/allocation/decider/NodeOverloadAwareAllocationDecider.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing.allocation.decider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.Setting.Property;
+
+import java.util.function.BiPredicate;
+
+/**
+ * This {@link NodeOverloadAwareAllocationDecider} controls shard over-allocation
+ * due to node failures or otherwise on the surviving nodes
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.provisioned_capacity: N
+ * </pre>
+ * <p>
+ * and prevent allocation on the surviving nodes of the under capacity cluster
+ * based on oveload factor defined as a percentage by
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.factor: X
+ * </pre>
+ */
+public class NodeOverloadAwareAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "overload_awareness";
+
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING =
```

Comment:
Today data nodes can be added/ removed from the cluster on demand and master balances the shards on the nodes present in the cluster at any time. But, now anytime nodes are added/ removed from the cluster, this setting has to be updated if it is a planned activity and the setting is turned on. 

During B/G this ProvisionedCapacity and node exclusion setting have to be set atomically otherwise it can leave shards unassigned or un-necessary movement. Thoughts?

Reply:
Yes they have to be set as a part of the same API call or this can be turned off while maintenance activity is underway and once the cluster is stable, it can be turned ON